### PR TITLE
feat(emp): financial product library to apply collateralization ratio transformation

### LINKED ISSUE #2160: Refactor `_getOraclePrice` and `_getCollateralRequirement` into external libraries, allow contract factories to have customizable constructor params
**Why**

Allows EMP deployers to modify how oracle price and collateral requirements are calculated, s.t. different types of financial contracts besides vanilla futures contracts can be created. 

Currently the EMP always uses the reported DVM price as its settlement price (relevant on settlements at expiry and disputes) and hardcodes its collateral.

But what if you wanted to create the following Expiring product:
- The contract requires sponsors to deposit 1 WETH in collateral and allows them to mint (and distribute) tokens that pay out that 1 WETH if, at expiry, ETHUSD is below a set strike. If ETHUSD is above that strike, the contract pays out a given dollar amount of WETH.
- Example: expiry is DEC 31. Strike is $400. Each token is backed by 1 WETH
- If ETHUSD < $400 at expiry, token is redeemed for 1 ETH.
- If ETHUSD >= $400 at expiry, token is redeemed for $400 worth of ETH, as determine by the oracle price returned at that time.

## `_getOraclePrice`
You could modify the existing EMP by changing the `_getOraclePrice()` method to:
```
oraclePrice = _getOraclePrice(expirationTimestamp);
// if price request is made before expiry, return 1. This means that we can keep
// the contract 100% collateralized with 1 WETH pre-expiry, and that disputes 
// pre-expiry are illogical (aka token is always backed by 1 WETH pre-expiry)
if (requestedTime < expirationTimestamp) {
     return 1;
}
if (oraclePrice < strike) {
  expiryPrice = 1;   // token expires to be worth 1 WETH
}
else {
  // token expires to be worth strike $ worth of WETH
  // eg if ETHUSD is $500 and strike is $400, token is
  // redeemable for 400/500 = 0.8 WETH
  expiryPrice = strike / _getOraclePrice(expirationTimestamp);
}
```

## `_getCollateralRequirement`
Currently EMP disputes enforce the following `requiredCollateral`:
```
 FixedPoint.Unsigned memory requiredCollateral = tokenRedemptionValue.mul(collateralRequirement);
```

But what's interesting about this contract is that the collateral required should decrease as ETHUSD goes up
```
ETHUSD = $350, you get 1 WETH
ETHUSD = $375, you get 1 WETH
ETHUSD = $400, you get 1 WETH
ETHUSD = $425, you get 0.941 WETH (worth $400)
ETHUSD = $450, you get 0.888 WETH (worth $400)
ETHUSD = $500, you get 0.8 WETH (worth $400)
etc.
```

Maybe we can change the collateral requirement dynamically?

## Flexible constructor params in the factory
This contract would require an additional `strike` constructor parameter.

**How**

Move these methods into a library that a deployer can separately deploy and link from within the EMPCreator.

Change the EMPCreator to allow flexible constructor params

**Difficulty Score [1-10]**

- 7: ~3 days

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

Issue #2160

Add the ability to transform contract collateral requirement based on an external library.

Follows on from PR https://github.com/UMAprotocol/protocol/pull/2185 which applied similar logic to the settlement price.

**Summary**

This PR introduces a function `transformCollateralRequirement` to the `FinancialProductLibrary` which enables the collateral requirement of a contract to be transformed by an external library. This is useful for contracts that might want to shift their collateral requirement as a function of price. An example of this is a structured note product with a defined strike price. If the price is above the strike then and additional collateral is redundant (and inefficient). 

**Details**

`transformCollateralRequirement` is applied to the `Liquidatable` contract by adding a `try`/`catch` wrapped method to transform the collateral requirement of the contract, if the library has the provided method. `Liquidatable` uses the `collateralRequirement` in exactly one place: computation of a position's correct collateralization at dispute settlement this can be seen here:
https://github.com/UMAprotocol/protocol/blob/bccaefe1ee4e187af03a91c3e84512d1d5353ba6/packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol#L530

This PR modifies this logic at dispute settlement time by changing the `_settle` function to use the transformed collateral requirement. 

**Issue(s)**

close #2160

**Bytecode impact**
Before this PR `ExpiringMultiPartyLib` had `595` remaining byte within the size limit. After this PR it has `350` bytes. Cutting it fine but still within the limits :)

### CODE REVIEW

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol [hunk 1]
```diff
@@ -504,6 +504,28 @@ contract Liquidatable is PricelessPositionManager {
...
     }
 
+    /**
+     * @notice Accessor method to calculate a transformed collateral requirement using the finanical product library
+      specified during contract deployment. If no library was provided then no modification to the collateral requirement is done.
+     * @param price input price used as an input to transform the collateral requirement.
+     * @return transformedCollateralRequirement collateral requirement with transformation applied to it.
+     * @dev This method should never revert.
+     */
+    function transformCollateralRequirement(FixedPoint.Unsigned memory price)
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:514]: Are you purposefully making this public?

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol [hunk 2]
```diff
@@ -526,8 +548,11 @@ contract Liquidatable is PricelessPositionManager {
         FixedPoint.Unsigned memory tokenRedemptionValue =
             liquidation.tokensOutstanding.mul(liquidation.settlementPrice);
 
-        // The required collateral is the value of the tokens in underlying * required collateral ratio.
-        FixedPoint.Unsigned memory requiredCollateral = tokenRedemptionValue.mul(collateralRequirement);
+        // The required collateral is the value of the tokens in underlying * required collateral ratio. The Transform
+        // Collateral requirement method applies a from the financial Product library to change the scaled the collateral
+        // requirement based on the settlement price. If no library was specified when deploying the emp then this makes no change.
+        FixedPoint.Unsigned memory requiredCollateral =
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:554]: What if the `sponsorCollateral < requiredCollateral` at liquidation time, but then at dispute time, the oracle price is high enough such that `requiredCollateral` decreases and `sponsorCollateral > requiredCollateral`, so the liquidator gets disputed successfully to their surprise.

This doesn't seem like a bug from the contract's point of view but is bad liquidator/disputer ux.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol [hunk 3]
```diff
@@ -504,6 +504,28 @@ contract Liquidatable is PricelessPositionManager {
...
+     * @return transformedCollateralRequirement collateral requirement with transformation applied to it.
+     * @dev This method should never revert.
+     */
+    function transformCollateralRequirement(FixedPoint.Unsigned memory price)
+        public
+        view
+        returns (FixedPoint.Unsigned memory)
+    {
+        if (address(financialProductLibrary) == address(0)) return collateralRequirement;
+        try financialProductLibrary.transformCollateralRequirement(price, collateralRequirement) returns (
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:520]: Have you considered adding a test that tests what happens when `financialProductLibrary. transformCollateralRequirement` reverts?  I did something similar when testing what would happen when the perpetual's `withdrawFundingRateFees()` method reverts [here](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/test/MockPerpetual.sol#L48)

**Review by nicholaspai — COMMENTED** [review]
This PR + the transformPrice PR are going to make the EMP very powerful, had some comments about some edge cases

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 1 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:514]: yes. it's a view function so no risk and it is needed for frontends to compute the current collateral requirement based on a given input price.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 3 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:520]: I have not tested such but can add a mock and a test to do it. will report back once it's added.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 2 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:554]: Hhhmm. I'm not sure I agree. This is different from the case in the previous library with the comment that @mrice32 brought up https://github.com/UMAprotocol/protocol/pull/2185#discussion_r525366736 that identified that for the transformation of price, you need to know the timestamp of the price request. This differs because we already know the `liquidation.settlementPrice` at the liquidation timestamp, as resolved by the DVM. The price requested to the DVM and the resolution there of are all around the liquidation timestamp. 

Therefore, we are always performing the transformation considering the timestamp _of the liquidation_. What this means is that the bot that ran the liquidation/dispute (or if this was liquidation/disputed from the front end) would first need to call `transformCollateralRequirement` to get their current value of the collateral requirement when computing if a position is correctly collateralized.

Does this make sense?

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 3 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:520]: I've added tests for reverting library to liquidatable & retroactively to the priceless position manager to verify this try/catch logic works as expected.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 2 above)
Comment (mrice32) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:554]: Yeah, @chrismaree, that makes sense to me.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol [hunk 4]
```diff
@@ -504,6 +504,28 @@ contract Liquidatable is PricelessPositionManager {
...
+     * @param price input price used as an input to transform the collateral requirement.
+     * @return transformedCollateralRequirement collateral requirement with transformation applied to it.
+     * @dev This method should never revert.
+     */
+    function transformCollateralRequirement(FixedPoint.Unsigned memory price)
+        public
+        view
+        returns (FixedPoint.Unsigned memory)
+    {
+        if (address(financialProductLibrary) == address(0)) return collateralRequirement;
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:519]: nit: if we're already using a try-catch here, would it make sense to _not_ do this explicit check and just let the try-catch handle the fallback?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/StructureNoteFinancialProductLibrary.sol [hunk 5]
```diff
@@ -75,4 +75,31 @@ contract StructuredNoteFinancialProductLibrary is FinancialProductLibrary, Ownab
...
+        // If the price is less than the strike than the original collateral requirement is used.
+        if (oraclePrice.isLessThan(strike)) {
+            return collateralRequirement;
+        } else {
+            // If the price is more than the strike then the collateral requirement is scaled by the strike. For example
+            // a strike of $400 and a CR of 1.2 would yield:
+            // ETHUSD = $350, payout is 1 WETH. CR is multiplied by 1. resulting CR = 1.2
+            // ETHUSD = $400, payout is 1 WETH. CR is multiplied by 1. resulting CR = 1.2
+            // ETHUSD = $425, payout is 0.941 WETH (worth $400). CR is multiplied by 0.941. resulting CR = 1.1292
+            // ETHUSD = $500, payout is 0.8 WETH (worth $400). CR multiplied by 0.8. resulting CR = 0.96
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/StructureNoteFinancialProductLibrary.sol:101]: very cool

**Review by mrice32 — APPROVED** [review]
LGTM! minor comments

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 4 above)
Comment (mrice32) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:519]: I also think the try-catch is subtly really cool here because it allows people to pass in financial product libraries that only implement the methods that they actually want to override.

Comment (chrismaree) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol]: this was intentional :) I initially did it the way you propose, (and the same is true with the price transformation) but I get an error as follows:
```
     Error: Transaction reverted: function call to a non-contract account
```

This occurs if the `financialProductLibrary` is set to an EOA, such as the address(0). This check would ensure that it would not revert at this point.

However, you did make me realize that I did not have a test for setting the `financialProductLibrary` to another EOA address! I tested this and indeed the same revert happens. 

I've updated the function to first check if the library is set to a contract. If it is not, then it does nothing. If it is, then it enters into the try-catch logic block.

I've also updated the previous library for price transformations to use the same logic.

PTAL.

File: packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol (hunk 2 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/expiring-multiparty/Liquidatable.sol:556]: yeah makes sense

**Review by nicholaspai — APPROVED** [review]
This was pretty easy to review because of the comments and thought out tests! Left some comments but LGTM after
